Add ecommerce tracking

JamesCGDS opened the pull request:
## What

This PR adds enhanced ecommerce tracking scripts that will gather data on pages such as search and push this data to the `dataLayer` where it can be accessed by Performance Analysts.

Caveat: on lines `90` and `91`, `internalLinksDomain` and `internalLinksDomainWithoutWww` are set directly because otherwise `isExternalLink` will return an error. This implementation is temporary and will be fixed following a refactor of the `ga4-link-tracker` code by @AshGDS (https://github.com/alphagov/govuk_publishing_components/pull/2961).

## Why

These changes are being made as part of our migration to GA4.

## Visual Changes

N/A

AshGDS commented on CHANGELOG.md:
```diff
@@ -7,6 +7,10 @@
   useful summary for people upgrading their application, not a replication
   of the commit log.
 
+## 30.5.3
```
This needs to be moved back to `## Unreleased` (I think a rebase causes this to happen as your change stays on the same line, but the heading changes)

JamesCGDS replied:
Done.

AshGDS commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+
+  Ga4EnhancedEcommerceTracker.prototype.populateEcommerceSchema = function (searchResultClicked = false, searchResult = null) {
+    // Limiting to 100 characters to avoid noise from extra longs search queries
+    // and to stop the size of the payload going over 8k limit.
```
Is this is a comment from UA, if so I think it should be removed (the 8k limit part)

JamesCGDS replied:
I think the comment should stay because I think it does relate to GA. Good spot re the typo, I've fixed that now.

AshGDS commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+    ecommerceObject.search_results.event_name = searchResultClicked && searchResult ? 'select_item' : 'view_item_list'
+    ecommerceObject.search_results.term = searchQuery
+    ecommerceObject.search_results.sort = variant
+    ecommerceObject.search_results.results = parseInt(this.getMetaTagContent('govuk:search-result-count'))
```
Is it OK that if the meta tag doesn't exist, `parseInt(null)` would return `NaN` and push that to the dataLayer? Should it fallback to `ecommerceRows.length` here if the meta tag doesn't exist?

JamesCGDS replied:
We can't use `ecommerceRows.length` because that doesn't return the total result count. As per our discussion, I've changed this function to get the total results from the `h2`.

AshGDS commented on spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js:
```diff
@@ -0,0 +1,295 @@
+
+    searchResults = document.createElement('div')
+    searchResults.innerHTML =
+      '<a data-ecommerce-path="/coronavirus" data-ecommerce-row="1" data-ecommerce-index="1" data-track-category="navFinderLinkClicked" data-track-action="Search.1" data-track-label="/coronavirus" data-track-options="{&quot;dimension28&quot;:20,&quot;dimension29&quot;:&quot;Coronavirus (COVID-19): guidance and support&quot;}" class="gem-c-document-list__item-title   govuk-link" href="/coronavirus ...
```
Might be worth removing the data values and the css class which aren't needed for these elements, so it's easier to read

JamesCGDS replied:
Agreed, I've removed those.

AshGDS reviewed (commented):
Thanks @JamesCGDS - I've added some comments :+1:

andysellick commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+    this.module = module
+  }
+
+  Ga4EnhancedEcommerceTracker.prototype.init = function (isNewPage = true) {
```
We don't normally pass parameters to the `init` function of a module like this, normally because Modules are called and initialised by the start Modules script. Instead, normal practice is to have another data attribute on the element in question and read it when the constructor is called, or in the init function. For example:

```HTML
<div data-module="Ga4EnhancedEcommerceTracker" data-is-new-page="false"></div>
```

```JavaScript
  function Ga4EnhancedEcommerceTracker (module) {
    this.module = module
    this.isNewPage = this.module.getAttribute('data-is-new-page') || true
  }
```

andysellick commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+    this.PIIRemover = new window.GOVUK.analyticsGA4.PIIRemover()
+    this.DEFAULT_LIST_TITLE = 'Site search results'
+
+    window.dataLayer = window.dataLayer || []
```
I think we should try to avoid initialising the dataLayer inside our code. Instead, I think we should be checking to see whether it exists as an early step, and backing out if it doesn't (on the assumption that GTM has failed to initialise for some reason, perhaps because cookie consent hasn't been given).

JamesCGDS replied:
Agreed, I've changed this to just be `if(window.dataLayer){ ... }`

andysellick commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+          event: 'page_view',
+          page_view: {
+            location: pageViewTracker.getLocation(),
+            referrer: pageViewTracker.getReferrer()
```
I'm a bit confused by this section of code - it looks like you're getting the page view code and using part of it to create a new page view in here? I feel like this could increase maintenance if we decide to change the data structure of page views. Instead, could you call the page view code directly?

```JavaScript
var pageViewTracker = window.GOVUK.analyticsGA4.analyticsModules.PageViewTracker
if (pageViewTracker) {
  pageViewTracker.init()
}
```

JamesCGDS replied:
As per our discussion earlier, I've implemented your above suggestion with a view to refactor the `PageViewTracker` code at a later date to accommodate scenarios where we only need to push a few select values of the `pageView` object to the `dataLayer` rather than the whole thing.

andysellick commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+    ecommerceObject.search_results.sort = variant
+    ecommerceObject.search_results.results = parseInt(this.getMetaTagContent('govuk:search-result-count'))
+    ecommerceObject.search_results.ecommerce = {}
+    ecommerceObject.search_results.ecommerce.items = []
```
Could this block be written out as an object instead of individual declarations? If it's possible I think it would help to make the structure of the object more plain.

```JavaScript
var ecommerceObject = {
  event: 'search_results',
  search_results: {
    event_name: searchResultClicked && searchResult ? 'select_item' : 'view_item_list',
    // etc.
  }
}
```

JamesCGDS replied:
Agreed, this is much easier to read.

andysellick commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+        item_id: searchResult.getAttribute('data-ecommerce-path'),
+        item_name: searchResult.textContent,
+        item_list_name: listTitle,
+        index: this.getIndex(searchResult, Array.prototype.indexOf.call(ecommerceRows, searchResult)) + startPosition
```
Can you explain what this line does?

JamesCGDS replied:
As per our discussion earlier, this code has been refactored to be more readable following a suggestion from @AshGDS.

andysellick commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+    return element.getAttribute('content');
+  }
+
+  Modules.Ga4EnhancedEcommerceTracker = Ga4EnhancedEcommerceTracker
```
Since this gets referred to in a few places, would it be cleaner to call this something shorter? Maybe just `Ga4EcommerceTracker`?

JamesCGDS replied:
Agreed. As per our discussion earlier, this will be addressed as part of a wider task to look at naming conventions.

andysellick commented on spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js:
```diff
@@ -0,0 +1,295 @@
+/* eslint-env jasmine */
+
+describe('Google Analytics ecommerce tracking', function () {
```
I'd consider merging the commit that adds this test with the commit that creates the code it refers to, since they're linked.

JamesCGDS replied:
Done.

andysellick commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -40,7 +40,7 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
       }
     }
-    
+
```
Could these linting fixes be merged with the commit that added these files?

JamesCGDS replied:
Done.

andysellick reviewed (commented):
Looks generally good, although it's a little hard to follow in places and I'm not sure it should be structured as one of our `GOVUK.Modules`. Let's have a chat and see what solution would be best.

JamesCGDS commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  Ga4EnhancedEcommerceTracker.prototype.trackSearchResults = function () {
+    var schema = this.populateEcommerceSchema(this.module)
```
Good spot!

JamesCGDS commented:
Thank you @andysellick and @AshGDS for your comments and input, it's much appreciated 👍  I've addressed those comments now, if you have any further feedback just give me a shout!

AshGDS commented:
Looks good to me but I will leave it for Andy to approve :+1:

andysellick commented on app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.js:
```diff
@@ -0,0 +1,144 @@
+        })
+
+        GOVUK.analyticsGA4.analyticsModules.Ga4LinkTracker.internalLinksDomain = 'www.gov.uk/'
+        GOVUK.analyticsGA4.analyticsModules.Ga4LinkTracker.internalLinksDomainWithoutWww = 'gov.uk/'
```
Hopefully a quick one. Can you recap why these need to be set? And will it have any detrimental effect on the rest of the code once these two lines have run?

JamesCGDS replied:
I’ve just pushed up a change that removes these. Initially, we were having problems without setting it but I’ve just tested this locally and everything works fine without them. The only place where we’d need to keep them is in the tests as the link tracker isn’t initialised there.

andysellick commented on spec/javascripts/govuk_publishing_components/analytics-ga4/ga4-enhanced-ecommerce-tracker.spec.js:
```diff
@@ -0,0 +1,296 @@
+    searchResults.innerHTML =
+      '<a data-ecommerce-path="/coronavirus" data-ecommerce-row="1" data-ecommerce-index="1" href="/coronavirus">Coronavirus (COVID-19): guidance and support</a>' +
+
+      '<a data-ecommerce-path="/guidance/travel-abroad-from-england-during-coronavirus-covid-19" data-ecommerce-row="1" data-ecommerce-index="2" href="/guidance/travel-abroad-from-england-during-coronavirus-covid-19">Travel abroad from England during coronavirus (COVID-19)</a>' +
```
Non blocking comment - it's good to base test data on real data but we can make our test data more readable by shortening things to the minimum i.e. the text and href of these links could be much shorter - as long as they're unique.

JamesCGDS replied:
Cool, I've just pushed up a change that will shorten those test links.

andysellick reviewed (approved):
Looks great, well done!
